dial_delay.go

```diff
@@ -0,0 +1,167 @@
+package swarm
+
+import (
+	"context"
+	"time"
+
+	ma "github.com/multiformats/go-multiaddr"
+	mafmt "github.com/whyrusleeping/mafmt"
+)
+
+const p_circuit = 290
+
+const numTiers = 2
+const tierDelay = 2 * time.Second
+
+var relay = mafmt.Or(mafmt.Base(p_circuit), mafmt.And(mafmt.IPFS, mafmt.Base(p_circuit)))
+
+// delayDialAddrs returns a address channel sorted by priority, pushing the
+// addresses with delay between them. The other channel can be used to trigger
+// sending more addresses in case all previous failed
+func delayDialAddrs(ctx context.Context, c <-chan ma.Multiaddr) (<-chan ma.Multiaddr, chan<- struct{}) {
+	out := make(chan ma.Multiaddr)
+	delay := time.NewTimer(tierDelay)
+	triggerNext := make(chan struct{}, 1)
+
+	go func() {
+		defer delay.Stop()
+		defer close(out)
+		var pending [numTiers][]ma.Multiaddr
+		lastTier := 0
+
+		// put enqueues the mutliaddr
+		put := func(addr ma.Multiaddr) {
+			tier := getTier(addr)
+			pending[tier] = append(pending[tier], addr)
+		}
+
+		// get gets the best (lowest tier) multiaddr available
+		get := func() (ma.Multiaddr, int) {
+			for i, tier := range pending[:] {
+				if len(tier) > 0 {
+					addr := tier[len(tier)-1]
+					tier[len(tier)-1] = nil
+					pending[i] = tier[:len(tier)-1]
+					return addr, i
+				}
+			}
+			return nil, -1
+		}
+
+	outer:
+		for {
+		fill:
+			for {
+				select {
+				case addr, ok := <-c:
+					if !ok {
+						break outer
+					}
+					put(addr)
+				default:
+					break fill
+				}
+			}
+
+			next, tier := get()
```

Comment:
Note that with the current setup we don't ever reach this point in this loop as the channel is closed by now. All sending will happen in the loop in L127. I'll add some tests later today to test that it works.